<a href="https://colab.research.google.com/github/AlejandroAgduelo22/Business-Analytics/blob/main/Modelo_Naive_Bayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Cargamos las librerias

In [ ]:
import numpy as np
import pandas as pd

from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix

1. Procedemos con la carga de los datos

In [ ]:
nxl='/content/1. BD2_CarCrash.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XDB.dropna() #Evitar celdas vacias
XDB.head(100)

#Creamos el DataMart
XD=XDB[['Weekday','ViolCat','ClearWeather','Month','Daylight']] #Variables Entrada
yd=np.array(XDB[['Highway']]) #Valor deseado - Aprenda el modelo


print("El numero de datos es:",len(XD))
len((np.where(yd==1))[0])/len(XD)

El numero de datos es: 6705


0.2714392244593587

2. Se procede con la configuración y evaluación del modelo Naive Bayes (Highway).

In [ ]:
mnb=GaussianNB()
mnb.fit(XD,yd)

#Vamos a analizar los resultados para el modelo

#Vamos a sacar las medias
u=mnb.theta_
print("Las medias de las variables son:\n",u)
#Vamos a sacar las desviaciones
sigma=np.sqrt(mnb.var_)
print("Las desviaciones de las variables son:\n",sigma)

#Procedemos con el análsis de las variaciones de las variables con LI y LS
#Limite Superior
LS=u+sigma
print("Los limites superiores para las variables son:\n",LS)
#Limite Inferior
LI=u-sigma
print("Los limites inferiores para las variables son:\n",LI)

#Vamos a evaluar el comportamiento del modelo

#Vamos a predecir los valores de accidentalidad en autopistas (Highway) del valor deseado
ydp=mnb.predict(XD) #Hizo el pronostico de accidentalidad en autopistas

#Vamos a evaluar el comportamiento del modelo
cm=confusion_matrix(yd,ydp)
print("La matriz de confusión es:\n",cm)

#Ya con la matriz de confusión, se procede a extraer los siguientes valores

VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]

Ex=(VP+VN)/len(XD)
print("La Exactitud del Modelo es:",Ex)

Ter=(FP+FN)/len(XD)
print("La Tasa de Error del Modelo es:",Ter)

Sen=VP/(FN+VP)
print("La sensibilidad del modelo es:",Sen)

Esp=VN/(VN+FP)
print("La Especifidad del modelo es:", Esp)

Pres=VP/(VP+FP)
print("La precision del modelo es",Pres)

PredNeg=VN/(VN+FN)
print("El valor de Predicción Negativo del modelo es",PredNeg)

#Vamos a evaluar si este accidente fue provocado en una autopista
XSC=[[7,1,1,3,0]]
ydsc=mnb.predict(XSC)
print("El accidente fue provocado en:",ydsc)




Las medias de las variables son:
 [[3.88945752 6.7778915  0.85261003 5.52548618 0.6814739 ]
 [3.82417582 4.31538462 0.79450549 5.47417582 0.70934066]]
Las desviaciones de las variables son:
 [[1.92831125 3.63475806 0.35449426 3.27379337 0.46590475]
 [1.82800803 2.36793707 0.40406253 3.26215279 0.45406663]]
Los limites superiores para las variables son:
 [[ 5.81776877 10.41264957  1.20710429  8.79927955  1.14737865]
 [ 5.65218385  6.68332168  1.19856802  8.73632861  1.16340729]]
Los limites inferiores para las variables son:
 [[1.96114628 3.14313344 0.49811577 2.25169281 0.21556915]
 [1.9961678  1.94744755 0.39044297 2.21202304 0.25527403]]
La matriz de confusión es:
 [[4560  325]
 [1563  257]]
La Exactitud del Modelo es: 0.7184190902311708
La Tasa de Error del Modelo es: 0.2815809097688292
La sensibilidad del modelo es: 0.1412087912087912
La Especifidad del modelo es: 0.9334698055271239
La precision del modelo es 0.4415807560137457
El valor de Predicción Negativo del modelo es 0.744732

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


**ANÁLISIS DE RESULTADOS**

Los resultados arrojados por el modelo propuesto frente al pronóstico de accidentalidad en autopistas, mostró que la precisión alcanzó un porcentaje del 44% lo que indica que de un total de 100 accidentes, el modelo tiende a equivocarse en 56%. Además,  la metrica de sensibilidad alcanzó un valor 14%  y una especifidad del 93% frente a la predicción de accidentalidad en autopistas. Es importante tener en cuenta que el infrome arroja resultados donde se evalua la accidentalidad en autopistas de los conductores teniendo en cuenta las disntas variables del contexto en donde se encuentre el conductor como:  Dia de la Semana, Nivel de Impacto (ViolCat), Mes, Cantidad de Luz Diurna, y Cielo Despejado, ademas, se debe considerar que todo ha sido analizado bajo el contexto de la ciudad de San Diego. Como información adicional, la base de datos posee un total de 6705 de registros de accidentes, donde se dividen en dos grupos de 4885 y 1820.
